tower-hedge/src/delay/mod.rs

```diff
@@ -0,0 +1,95 @@
+use futures::{Async, Future, Poll};
+use tower_service::Service;
+
+mod error;
+use self::error::Error;
+
+use std::time::Duration;
+
+/// A policy which specifies how long each request should be delayed for.
+pub trait Policy<Request> {
+    fn delay(&self, req: &Request) -> Duration;
+}
+
+/// A middleware which delays sending the request to the underlying service
+/// for an amount of time specified by the policy.
+pub struct Delay<P, S> {
+    policy: P,
+    service: S,
+}
+
+enum State<Request, F> {
+    Delaying(tokio_timer::Delay, Option<Request>),
+    Called(F),
+}
+
+pub struct ResponseFuture<Request, S, F> {
+    service: S,
+    state: State<Request, F>,
+}
+
+impl<P, S> Delay<P, S> {
+    pub fn new<Request>(policy: P, service: S) -> Self
+    where
+        P: Policy<Request>,
+        S: Service<Request> + Clone,
+        S::Error: Into<super::Error>,
+    {
+        Delay { policy, service }
+    }
+}
+
+impl<Request, P, S> Service<Request> for Delay<P, S>
+where
+    P: Policy<Request>,
+    S: Service<Request> + Clone,
+    S::Error: Into<super::Error>,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = ResponseFuture<Request, S, S::Future>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.service
+            .poll_ready()
+            .map_err(|e| Error::ServiceError(e.into()))
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        let deadline = tokio_timer::clock::now() + self.policy.delay(&request);
+        ResponseFuture {
+            service: self.service.clone(),
```

Comment:
The instance that `poll_ready` was called on needs to be the one passed to `ResponseFuture`. You probably want to call `clone`, then use `mem::swap` to place the cloned service onto `self`.

If tests pass using an up-to-date version of `tower_test::mock`, then there is a bug in the mock implementation. It should catch this case.